proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,733 @@
+relation as it is.
+
+We add a new relation ``refersTo`` which relates qualified names to
+entities in a scope. [#]_
+
+.. [#] By analogy to programming languages: the bindings in scope are like an environment,
+   simple names are like variables, qualified names are like expressions, and ``refersTo``
+   is ``evalExpr``.
+
+::
+
+   inScope :: Rel Module Name Entity
+   exports :: Rel Module Name Entity
+   refersTo :: Rel Module QualName Entity
+
+``refersTo`` works as follows:
+
+- If the qualified name is a (possibly dotted) simple name, then 
+  it refers to whatever that simple name is bound to in the current scope.
+- If the qualified name has a qualifier, then we work out what
+  the qualifier refers to in the current scope. If it refers to a module,
+  then we work out what the simple name refers to in the exported environment of
+  that module.
+
+That is, we interpret ``A.b`` by first working out what ``A`` refers to, as a
+module, and then looking inside it to find ``b``. ``A`` may itself be
+qualified (since modules may export module bindings, just like any other
+bindings), hence the need for the recursive lookup.
+
+Imports
+^^^^^^^
+
+Normal imports can only appear at the top level of a file. The target of a 
+normal import is a *dotted* name, which is resolved using the
+current rules for locating external Haskell modules.
+
+Importing a module brings all the exported bindings of the target module
+into scope, as described in [hsmods]_. It also brings
+the target module into scope bound to its declared simple name (which may be
+a dotted name if it is a file module).
+
+If a module is imported ``qualified`` then the exported bindings of the target
+module are not brought into scope.
+
+If a module is imported ``as`` a simple name, then the target module is brought
+into scope bound to that simple name instead.
+
+Module export specifiers
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The behaviour of the ``module`` export specifiers are changed as follows.
+
+- ``module M`` *changes* to exports the name ``M`` as a module binding.
+- ``module M(a, b, .. c)`` exports ``M`` and the listed in-scope bindings.
+- ``module M(..)`` exports ``M`` and all of its exported bindings. [#]_
+
+.. [#] See `Do we need additional syntax for re-exporting all bindings from a module?`_ for more discussion.
+
+Backwards compatibility extensions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Flat module exports
+^^^^^^^^^^^^^^^^^^^
+
+The extension ``FlatModuleExports`` is added and enabled by default. It has the effect
+that the ``module M`` export specifier has the behaviour given in [hs2010]_. [#]_
+
+.. [#] See `Flat module exports - reasoning`_ for reasoning.
+
+Import module merging
+^^^^^^^^^^^^^^^^^^^^^
+
+The extension ``ImportModuleMerging`` is added and enabled by default. It has the effect that
+when multiple modules are imported ``as`` the same name, we merge the modules. [#]_
```

Comment:
It is sadly possible to have

```
import A
import B as A
```

The first `import` does not import `A` `as` anything, so this sentence might want to be generalized a bit, if I suppose your intended meaning correctly.

Reply:
Sigh. Yes, it sounds like this needs to cover all cases where module names are introduced as part of import statements.